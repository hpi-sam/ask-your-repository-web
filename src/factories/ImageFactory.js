// @flow
import _ from 'lodash';
import faker from 'faker';
import type { Image } from '../models/Image';

class ImageFactory {
  static createDummyImage(tagAmount: number = 3): Image {
    return {
      id: faker.random.number(),
      url: faker.image.dataUri(),
      tags: _.times(tagAmount, () => faker.lorem.slug()),
    };
  }

  static createStaticDummyImage(): Image {
    return {
      id: 3826,
      url: 'data:image/svg+xml;charset=UTF-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20version%3D%221.1%22%20baseProfile%3D%22full%22%20width%3D%22undefined%22%20height%3D%22undefined%22%3E%20%3Crect%20width%3D%22100%25%22%20height%3D%22100%25%22%20fill%3D%22grey%22%2F%3E%20%20%3Ctext%20x%3D%220%22%20y%3D%2220%22%20font-size%3D%2220%22%20text-anchor%3D%22start%22%20fill%3D%22white%22%3Eundefinedxundefined%3C%2Ftext%3E%20%3C%2Fsvg%3E',
      tags: ['nemo-dolorum-iusto', 'aut-quia-atque', 'et-velit-dolorem'],
    };
  }
}

export default ImageFactory;
