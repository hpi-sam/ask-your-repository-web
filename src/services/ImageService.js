// @flow
import humps from 'humps';
import api from '../config/api';
import type { Image } from '../models/Image';
import type { Tag } from '../models/Tag';

type ListParams = {
  teamId: string,
  offset?: number,
  limit?: number,
  search?: string,
};

class ImageService {
  static async get(id: string): Promise<Image> {
    const response = await api.get(`/images/${id}`);
    const image: any = humps.camelizeKeys(response.data);
    return image;
  }

  static async list(params: ListParams): Promise<Image[]> {
    const response = await api.get('/images', {
      params: humps.decamelizeKeys(params),
    });
    const images: any = humps.camelizeKeys(response.data.images);
    return images;
  }

  static async related(id: string, limit: number = 5) {
    const response = await api.get(`/images/${id}/related`, {
      params: { limit },
    });
    const images: any = humps.camelizeKeys(response.data.images);
    return images;
  }

  static async create(formData: FormData): Promise<Image> {
    const headers = { 'Content-Type': 'multipart/form-data' };
    const response = await api.post('/images', formData, { headers });
    const image: any = humps.camelizeKeys(response.data);
    return image;
  }

  static async addTags(imageId: string, tags: Array<Tag>): Promise<Image> {
    const response = await api.post(`/images/${imageId}/tags`, { tags });
    const image: any = humps.camelizeKeys(response.data);
    return image;
  }

  static async patch(imageId: string, data: Object) {
    await api.patch(`/images/${imageId}`, data);
  }

  static async delete(imageId: string) {
    await api.delete(`images/${imageId}`);
  }

  static async patchMany(images: Array<{ id: string, tags: Array<Tag> }>) {
    await api.patch('/images', { artifacts: images });
  }
}

export default ImageService;
