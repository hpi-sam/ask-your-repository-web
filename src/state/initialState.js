// @flow

// This file is only for state related testing

import type { AppState } from './AppState';
import { initialState as presentation } from './presentation/presentation.reducer';
import { initialState as teamSidebar } from './team_sidebar/teamSidebar.reducer';
import { initialState as presentationMode } from './presentation_mode/presentationMode.reducer';
import { initialState as activeTeam } from './active_team/activeTeam.reducer';
import { initialState as auth } from './auth/auth.reducer';

const initialState: AppState = {
  flash: { messages: [] },
  presentation,
  activeTeam,
  teamSidebar,
  presentationMode,
  auth,
};

export default initialState;
