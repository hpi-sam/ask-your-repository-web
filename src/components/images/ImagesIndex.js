// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import type { Location } from 'react-router';
import { Route } from 'react-router-dom';
import { IoIosCloudOutline, IoIosSearch } from 'react-icons/io';
import qs from 'qs';
import InfinityScroll from 'react-infinite-scroller';
import Gallery from './gallery/Gallery';
import ImageService from '../../services/ImageService';
import ImageDecorator from './gallery/ImageDecorator';
import ActivityIndicator from '../utility/ActivityIndicator';
import type { Team } from '../../models/Team';
import type { Image as APIImage } from '../../models/Image';
import type { Image } from './gallery/ImageDecorator';
import type { AppState } from '../../state/AppState';
import ListEmpty from '../utility/ListEmpty';
import { ButtonLink } from '../utility/buttons';
import ImageDetailsModal from './ImageDetailsModal';
import './ImagesIndex.scss';

type Props = {
  activeTeam: ?Team,
  location: Location,
};

type State = {
  images: Array<Image>,
  offset: number,
  endReached: boolean,
  isLoadingInitially: boolean,
};

export const limit = 20;

class ImagesIndex extends Component<Props, State> {
  defaultState = {
    images: [],
    offset: 0,
    endReached: false,
    isLoadingInitially: true,
  };

  constructor(props: Props) {
    super(props);

    this.state = this.defaultState;
  }

  async componentDidMount() {
    await this.loadMoreImages();
  }

  componentDidUpdate(prevProps: Props) {
    if (this.props.activeTeam !== prevProps.activeTeam) {
      this.reloadImages();
    }
    if (this.props.location.search !== prevProps.location.search) {
      this.reloadImages();
    }
  }

  getSearchString = (querystring) => {
    const { search } = qs.parse(querystring, { ignoreQueryPrefix: true });
    return search;
  };

  reloadImages = () => {
    this.setState(this.defaultState, this.loadMoreImages);
  };

  loadMoreImages = async () => {
    if (this.state.endReached) return;

    const currentOffset = this.state.offset;
    this.increaseOffset();

    const { activeTeam } = this.props;
    if (!activeTeam) return;

    const search = this.getSearchString(this.props.location.search);

    try {
      const params = {
        teamId: activeTeam.id,
        offset: currentOffset,
        limit,
        search,
      };
      const images: APIImage[] = await ImageService.list(params);
      const filteredImages = images
        .filter(image => image.score === undefined || image.score > 0);
      this.receiveImages(filteredImages);
    } catch (error) {
      // TODO: Handle error
    }

    this.setState({ isLoadingInitially: false });
  };

  deleteImage = (id: string) => {
    this.setState(state => ({
      images: state.images.filter(image => image.id !== id),
    }));
  };

  increaseOffset() {
    const { offset } = this.state;
    this.setState({ offset: offset + limit });
  }

  receiveImages(fetchedImages: APIImage[]) {
    const decoratedImages = fetchedImages.map(
      image => ImageDecorator.decorateImage(image, this.deleteImage),
    );
    this.setState(state => ({ images: [...state.images, ...decoratedImages] }));

    if (fetchedImages.length === 0) {
      this.setState({ endReached: true });
    }
  }

  render() {
    const { images, isLoadingInitially } = this.state;
    const search = this.getSearchString(this.props.location.search);

    if (isLoadingInitially) {
      return (
        <ActivityIndicator
          centered
          text="Loading images..."
        />
      );
    }

    if (images.length === 0 && search) {
      return (
        <ListEmpty>
          <IoIosSearch />
          <span>
            We did not find any images
            <br />
            for your search query :(
          </span>
        </ListEmpty>
      );
    }

    if (images.length === 0) {
      return (
        <ListEmpty>
          <IoIosCloudOutline />
          <span>
            Your team has no images yet.
            <br />
            Upload your first one!
          </span>
          <ButtonLink to="/upload">
            Upload Image
          </ButtonLink>
        </ListEmpty>
      );
    }

    return (
      <div className="ImagesIndex">
        <Route path="/images/:id" component={ImageDetailsModal} />
        <InfinityScroll
          initialLoad={false}
          hasMore={!this.state.endReached}
          loadMore={this.loadMoreImages}
        >
          <Gallery images={images} />
        </InfinityScroll>
      </div>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  activeTeam: state.activeTeam,
});

export default connect(mapStateToProps)(ImagesIndex);
