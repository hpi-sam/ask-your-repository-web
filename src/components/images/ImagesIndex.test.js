// @flow
import React from 'react';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import ImagesIndex, { limit } from './ImagesIndex';
import ImageService from '../../services/ImageService';
import ImageFactory from '../../factories/ImageFactory';
import TeamFactory from '../../factories/TeamFactory';
import initialState from '../../state/initialState';

const mockStore = configureMockStore();

jest.mock('../../services/ImageService');

describe('<ImagesIndex />', () => {
  let wrapper;
  let store;

  const state = {
    ...initialState,
    activeTeam: TeamFactory.createStaticDummyTeam(),
  };

  beforeAll(() => {
    ImageService.list.mockImplementation(() => Promise.resolve([]));
  });

  beforeEach(() => {
    store = mockStore(state);
    const location = { search: ' ' };
    wrapper = mount((
      <Provider store={store}>
        <MemoryRouter>
          <ImagesIndex location={location} />
        </MemoryRouter>
      </Provider>
    )).find('ImagesIndex');
  });

  afterEach(() => {
    ImageService.list.mockClear();
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('server responding with a non empty array', () => {
    beforeAll(() => {
      const images = [ImageFactory.createDummyImage()];
      ImageService.list.mockImplementation(() => Promise.resolve(images));
    });

    it('should increase the offset by the limit', async () => {
      expect(wrapper.state().offset).toEqual(limit);
    });
  });

  it('should not call image service when end is reached', async () => {
    const calls = ImageService.list.mock.calls.length;
    wrapper.setState({ endReached: true });
    await wrapper.instance().loadMoreImages();
    expect(ImageService.list).toHaveBeenCalledTimes(calls);
  });

  describe('server responding with empty array', () => {
    beforeAll(() => {
      ImageService.list.mockImplementation(() => Promise.resolve([]));
    });

    it('should set state to end reached', async () => {
      expect(wrapper.state().endReached).toEqual(true);
    });
  });
});
