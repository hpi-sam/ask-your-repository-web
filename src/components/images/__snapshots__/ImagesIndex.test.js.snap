// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImagesIndex /> renders correctly 1`] = `
<ImagesIndex
  activeTeam={
    Object {
      "id": "2dedf601-f0bf-4290-ac3c-f5761234d068",
      "joinKey": "bbaf37e7-6d47-45e3-8538-a20a48920ed7",
      "members": Array [
        Object {
          "email": "test@example.com",
          "hasPassword": true,
          "id": "f1bc8f40-7698-4df7-b410-efac1f1a03d2",
          "username": "test_user",
        },
      ],
      "name": "odio expedita aut",
    }
  }
  dispatch={[Function]}
  location={
    Object {
      "search": " ",
    }
  }
>
  <ActivityIndicator
    centered={true}
    className=""
    text="Loading images..."
  >
    <div
      className="ActivityIndicator -centered"
    >
      <MdCropSquare
        className="ActivityIndicator__spinner"
      >
        <IconBase
          attr={
            Object {
              "viewBox": "0 0 24 24",
            }
          }
          className="ActivityIndicator__spinner"
        >
          <svg
            className="ActivityIndicator__spinner"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 4H6c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 14H6V6h12v12z"
              key="0"
            />
          </svg>
        </IconBase>
      </MdCropSquare>
      <div
        className="ActivityIndicator__text"
      >
        Loading images...
      </div>
    </div>
  </ActivityIndicator>
</ImagesIndex>
`;
